Iox 1032 new error handling api (#1902)
eclipse-iceoryx/iceoryx: doc/design/error_reporting.md

technical limitation from macro usage (required for location) as it is not allowed to have empty variadic
macros.

### Report an Error

To report an error the error code (later to be extended to error objects) has to be provided

```cpp
IOX_REPORT(Code::OutOfMemory, RUNTIME_ERROR);
```

This reports a `OutOfMemory` error and continues execution.

### Report a Fatal Error

Similarly

```cpp
IOX_REPORT_FATAL(Code::OutOfMemory);
```

reports a fatal error that aborts execution after the custom implementation specific handler is invoked.

Decoupling the error and its category is intentional, as e.g. an `OutOfMemory` error may not always be
fatal. It may become fatal after it is propagated further along the call stack. 
Propagation is possible by various means, e.g. return codes, monadic types or even exceptions (that
must be caught before reporting the error elsewhere again).

### Conditionally Report an Error

Conditionally reporting an error if some condition holds is useful for more compact error reporting.

```cpp
int x;
// ...
IOX_REPORT_IF(x<0, Code::OutOfBounds, RUNTIME_ERROR)
```

### Assert That a Condition Holds

> Similarly we can conditionally check whether a condition does not hold and report a fatal error in
> the case that it does not

Reviewer: ```suggestion
Similarly we can conditionally check whether a condition does hold and report a fatal error in
the case that it does not
```
Otherwise it is like a double negation.

This should be used for conditions that may not hold on the correct path, e.g. for error cases.
It should not be used for assumptions that have to be true in correct code 
(use `IOX_ASSUME` or `IOX_PRECONDITION` for this).

Note it that no condition can generally be enforced in the sense that it must be true and no checking is required.

## Using the API to Check Contracts and Assumptions

The following checks can be disabled and are intended to increase safety during incorrect
use, specifically detect incorrect use at runtime.

This means they should only be used for checks that are not needed in correct code (i.e. defensive
programming).

If these checks are disabled, there is no overhead in the code, i.e. no checking or reporting takes place.

### Checking Preconditions

A precondition check

```cpp
int f(int x)
{
IOX_PRECONDITION(x>=0, "precondition violation message")

// ...
}
```

is used to verify assumptions **BEFORE** any logic in the function body is executed. Technically copy
constructors may run before any condition can be checked, and there is also the possibility of
reordering if the following code does not depend on the condition at all. This is a limitation of

> the language and cannot be avoided.

Author: Yes but the compiler may move code below the precondition check (without observable effects including side effects) before the precondition check unless I miss something. So it is not a problem, but still not as it should work for "perfect precondition checking", namely just before the function is called.

reordering if the following code does not depend on the condition at all. 
This is not a problem since any reordering is not allowed to affect the observable result. 
Specifically it cannot affect the value of the precondition itself as this would change the
observable behaviour.

In case of violation, the violation and a (potentially empty) message are forwarded to the backend,
panic is invoked and execution stops.

The verification can be optionally disabled, and hence this also documents assumptions of the
function itself.

### Checking Assumptions

Checking assumptions is similar to checking preconditions, but can happen anywhere in the code.

```cpp
int f(int x)
{
    // some computation
    int y = g(x);

    IOX_ASSUME(y>=0, "assumption violation message");
    // proceed assuming that y>=0 holds
}
```

These serve as documentation of assumptions that should hold at this point in the code before the
next statement and can be used e.g. to check for out of bounds accesses. It can also be used to
check postconditions.

It should not be used at the start of a function body and instead replaced with a precondition check
in this case.

> ### Marking Unreachable Code

Reviewer 2: This should not be under a section which states that it can be disabled at compile time, like `## Using the API to Check Contracts and Assumptions` indicates

Author: Put into its own section (reduced header indent level).

```
if there is no message.

### Marking Unreachable Code

It is also possible to explcitly state that code is supposed to be unreachable.
```cpp
if(condition) {
    // Reachable code that does something
    // This also implies that it is assumed that the condition cannot be false.
} else {
    IOX_UNREACHABLE();
    // Code here should be dead, otherwise it is a bug
    // There should ideally be no dead code, but there are exceptions.
}
```

> If checking for unreachable code at runtime is enabled and `IOX_UNREACHABLE` is reached, 

Author: It is not possible anymore. The documentation is inconsistent and will be changed.

> If checking for unreachable code at runtime is enabled and `IOX_UNREACHABLE` is reached, 

Author: Changed.

> If checking for unreachable code at runtime is enabled and `IOX_UNREACHABLE` is reached, 

Author: Done

> If checking for unreachable code at runtime is enabled and `IOX_UNREACHABLE` is reached, 

Author: @FerdinandSpitzschnueffler It should not anymore, but maybe I missed it while reading the paragraph again.

> }

Reviewer 2: A better example would be an exhausting switch statement with a return in each case and an `IOX_UNREACHABLE` after the switch statement. This is a real world use case since one can cast any number to an enum class.

Author: I added the example but did not remove the previous one because I do not agree it is better, it is just another use case.

`panic` will be invoked and the program aborts. Stating that specific code cannot be reached is
a specific assumption and any violation is considered a bug (hence checking can be optionally
disabled).

This has advantages for test coverage as the compiler and other tools that rely on the compiler (say
for coverage) are aware of the `noreturn` guarantee of `IOX_UNREACHABLE`.

> As a consequnece, branches with `IOX_UNREACHABLE` do not necessarily lead to a return statement.

Reviewer: ```suggestion
As a consequence, branches with `IOX_UNREACHABLE` do not necessarily lead to a return statement.
```

against bug conditions such as precodition violations. As the latter should not happen in correct
code, these can be disabled.

## Examples

The following examples show how non-fatal and fatal errors can be signaled.

## Recoverable Errors

The default version only supports error codes.
The following assumes there exists a `RUNTIME_ERROR` category.

```cpp
enum class Code {
    SomeError
};

expected<int, Code> algorithm(int x)
{
    if(errorCondition(x))
    {
        IOX_REPORT(SomeError, RUNTIME_ERROR);
        // control flow continues and the error is propagated to the caller

>         return error<Code>(SomeError);

Reviewer 2: ```suggestion
        return err(SomeError);
```
There is an open PR which changes the API

    {
        IOX_REPORT(SomeError, RUNTIME_ERROR);
        // control flow continues and the error is propagated to the caller
        return error<Code>(SomeError);
    }
    return 42;
}

expected<int, E> identity() {
    
    auto result = algorithm(73);

    if(result.has_error())
    {
        // transform the error to E and propagate it
        return into<error<E>>(result.get_error());
    }
    
    // no error, return identity
    return *result;
};
```

This is similar to exception handling without the convenience of propagation.
While this shows the use with `expected`, it can be used with any error return type, 
for example the error code itself.

A generalization allows to report more complex error types directly. This requires a corresponding
custom implementation.

```cpp
expected<int, Code> algorithm(int x)
{
    if(errorCondition(x))
    {
        // create an exception like custom error
        auto e = error<CustomError>(SomeError, "additional error info");
        // report e directly
        IOX_REPORT_FATAL(e, RUNTIME_ERROR);
        return e;
    }
    return 42;
}

expected<int, E> identity()
{

>     auto result = algorithm();

Author: Why? I mean in theory it could be failing due to side effect but I will add an argument. 

Reminds me of a Monty Python sketch actually. I like to have an argument...

>     auto result = algorithm();

Author: The naming `identity` was off as well, the example changed but the naming did not. Updated.

        return into<error<E>>(result.get_error());
    }
    // no error, return result

>     return *result;

Author: But we do not as of now as far as I am aware so I changed it as suggested.

        return into<error<E>>(result.get_error());
    }
    // no error, return identity
    return *result;
};
```

## Non-recoverable Errors

Non-recoverable errors should generally not be used in combination with return codes or other error
types, since the control flow does not return from a fatal error.

```cpp
int algorithm(int x)
{
    if(errorCondition(x))
    {
        IOX_REPORT_FATAL(SomeError);
        // does not return, so no return statement is required
    }
    return 42;
}

int identity()
{

>     auto result = algorithm();

Reviewer: Same here.

Author: Same as above. Will change.

    // if a result was returned, we know that no error has occured
    return result;
};
```

Alternatively the shorthand version can be used

```cpp
int algorithm(int x)
{
    // require that the condition holds or raise a fatal error
    IOX_REQUIRE(!errorCondition(x), SomeError);
    return 42;
}
```

A generalization to error types other than codes is possible with a corresponding 
backend implementation.

## Structure

### Basics

> The overall implementation concepts allows customizationof the implementation level and a provides 

Reviewer: ```suggestion
The overall implementation concepts allow customization of the implementation level and provide
```

a default implementation.

Everything related to error reporting is located in the corresponding folder `error_reporting`.
Since the main API is stateless, there is no need for classes. Everything directly in this folder
(i.e. not in a subfolder) is not supposed to be changed.

These are

> 1. `api.h`: the reporting API to be used

Author: I will change the name. It was just to make the distinction between (public) API and auxiliary code clear.

> 1. `api.h`: the reporting API to be used

Author: Changed as requested.

2. `configuration.hpp`: the default configuration (compile time flags)
3. `error_forwarding.hpp`: forwarding to the custom implementation
4. `error_kind.hpp`: mandatory error categories
5. `error_logging.hpp`: logging related definitions
6. `location.hpp`: source location related definitions

> 7. `types.hpp`: auxiliary types

Author: Added.

> All definitions have to reside in `iox::err`, which is considered a private (detail) namespace 

Author: I see an renaming of my namespace on the horizon. OK, I think about a name. No big deal. Will be done after a complete review.

> All definitions have to reside in `iox::err`, which is considered a private (detail) namespace 

Author: Renamed `iox::err` to `iox::er` as discussed offline.

All the files focus on singular aspects to allow fine-grained inclusion.
All definitions have to reside in `iox::err`, which is considered a private (detail) namespace 
for everything related to error reporting. Since the API uses macros, it has no namespace itself.

### Custom Implementation

A specific custom implementation may depend on any of them and has to implement an error reporting
interface `error_reporting.hpp`. 

Apart from implementing the error reporting interface, a custom implementation does not have to follow 
a specific structure. However, it cannot depend on anything that intends to use the error 

> reporting itself. It can override or extend some definitions and it is encouraged to use the same 

Author: Yes. This is a problem I also encountered when plying with support for expected.

file names as in the mandatory absics. For example `custom/error_kind.hpp` specifies additional 
error kinds (apart from the mandatory fatal errors).

The main purpose of thec ustom implementation is to define the actions to take for each error.

### Default Implementation

The default implementation in `custom/default` allows switching between a `DefaultHandler` 
and a `TestHandler` at runtime.

> The latter is used in testing to verify that an error occured when the it is expected.

Reviewer: ```suggestion
The latter is used in testing to verify that an error occurred when it is expected.
```

The `DefaultHandler` is deployed outside of tests and provides minimal logging information.

### Testing

All testing related definitions are located in `iceoryx_hoofs/testing/error_reporting`.
These are the definition of `TestHandler` in `test_error_handler.hpp` and auxiliary functions in
`test_support.hpp` to be used in tests to verify errors. The latter can be extended as required.

### Custom Implementation

Extension of existing definitions is possible by either changing the default implementation 

> or providing an additional custom implementation and ensure that it is used by all modules.

Reviewer: This should be combined with the `Custom Implementation` section above.

### Modules

There must be a single point where all modules are defined to ensure they use unique ids and use the
same custom implementation. Currently this happens in the `modules` folder but is work in progress to be
completed during integration of error reporting.

There is `modules/hoofs/error_reporting.hpp` that defines all the errors and custom implementation 
used by `iceoryx_hoofs`. The `api.hpp` is included there to make it easy to use the custom error 
reporting in any iceoryx hoofs file by including `modules/hoofs/error_reporting.hpp`.

Replacing the previous error handling is supposed to happen by

1. Adapting the error definitions for `iceoryx_hoofs` in `modules/hoofs/errors.hpp`
2. Introducing a similar folder structure for `iceoryx_posh`

> 3. Replacing occurences of the previous error handler call (including `cxx::Expects`)

Reviewer: ```suggestion
3. Replacing occurrences of the previous error handler call (including `cxx::Expects`)
```

    // proceed assuming that y>=0 holds
}
```

These serve as documentation of assumptions that should hold at this point in the code before the
next statement and can be used e.g. to check for out-of-bounds accesses. It can also be used to
check postconditions.

It should not be used at the start of a function body and instead replaced with a precondition check
in this case.

## Marking Unreachable Code

It is also possible to explicitly state that code is supposed to be unreachable.
This type of check is always active but does not incur a performance penalty unless the code is
reached.

```cpp
if(condition) {
    // Reachable code that does something
    // This also implies that it is assumed that the condition cannot be false.
} else {
    IOX_UNREACHABLE();
    // Code here should be dead, otherwise it is a bug
    // There should ideally be no dead code, but there are exceptions.
}
```

Another use case is to convey intention by marking exhaustive `switch` statements.

```cpp
enum class Color {
    Red,
    Blue
};

int handleColor(Color color) {

    switch(color) {
        case Color:Red { return handleRed(); }

        case Color:Blue { return handleBlue(); }

>     }

Author: Yes ... I am not hat my best today. Fixed.

}
```

Another use case is to convey intention by marking exhaustive `switch` statements.

```cpp
enum class Color {
    Red,
    Blue
};

void handleColor(Color color) {

    switch(color) {
        case Color:Red { handleRed(); }

>         case Color:Blue { handleBlue(); }

Author: True, oversight on my part. I should not do dry coding, even if it it is trivial.
